Loading the weights from old model and continuing training apparently broken

Scenario:

Model trained for 100 epochs and got to 50% accuracy with validation graphs looking good (yay). Loaded the model weights for another go with nothing changed (cloned job and loaded weights): flat line under 10 percent for 22 epochs. I've tried continuing many other jobs and it's always similar. Model weights apparently get loaded, but results are nowhere near where they were after the first run.

(3.1. dev)

> I thought the renaming of the last layer was to prevent an error when retraining a new network with a different number of outputs but when I tried to train on a 2-class dataset after training on a 10-class dataset this worked seamlessly.

You're right - that's why we do it. When I comment out those lines and try to retrain on a dataset with a different number of labels, I get error messages like this:
```
2016-01-21 08:58:36 [20160121-085831-93e4] [ERROR] Train Caffe Model: Cannot copy param 0 weights from layer 'fc8'; shape mismatch.  Source param shape is 2 4096 (8192); target param shape is 20 4096 (81920). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer.
2016-01-21 08:58:44 [20160121-085838-0a15] [ERROR] Train Caffe Model: Cannot copy param 0 weights from layer 'fc8'; shape mismatch.  Source param shape is 20 4096 (81920); target param shape is 2 4096 (8192). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer.

```

--- a/digits/frameworks/caffe_framework.py
+++ b/digits/frameworks/caffe_framework.py
@@ -86,17 +86,20 @@ class CaffeFramework(Framework):
         return network
 
     @override
-    def get_network_from_previous(self, previous_network):
+    def get_network_from_previous(self, previous_network, use_same_dataset):
         """
         return new instance of network from previous network
         """
         network = caffe_pb2.NetParameter()
         network.CopyFrom(previous_network)
-        # Rename the final layer
-        # XXX making some assumptions about network architecture here
-        ip_layers = [l for l in network.layer if l.type == 'InnerProduct']
-        if len(ip_layers) > 0:
-            ip_layers[-1].name = '%s_retrain' % ip_layers[-1].name
+
+        if not use_same_dataset:
+            # Rename the final layer
+            # XXX making some assumptions about network architecture here
+            ip_layers = [l for l in network.layer if l.type == 'InnerProduct']
+            if len(ip_layers) > 0:
+                ip_layers[-1].name = '%s_retrain' % ip_layers[-1].name
+
         return network
 
     @override

--- a/digits/frameworks/framework.py
+++ b/digits/frameworks/framework.py
@@ -48,7 +48,7 @@ class Framework(object):
         """
         raise NotImplementedError('Please implement me')
 
-    def get_network_from_previous(self, previous_network):
+    def get_network_from_previous(self, previous_network, use_same_dataset):
         """
         return new instance of network from previous network
         """

--- a/digits/frameworks/torch_framework.py
+++ b/digits/frameworks/torch_framework.py
@@ -76,11 +76,18 @@ class TorchFramework(Framework):
         return network_desc
 
     @override
-    def get_network_from_previous(self, previous_network):
+    def get_network_from_previous(self, previous_network, use_same_dataset):
         """
         return new instance of network from previous network
         """
-        # return the same string
+        # note: use_same_dataset is ignored here because for Torch, DIGITS
+        # does not change the number of outputs of the last linear layer
+        # to match the number of classes in the case of a classification
+        # network. In order to write a flexible network description that
+        # accounts for the number of classes, the `nClasses` external
+        # parameter must be used, see documentation.
+
+        # return the same network description
         return previous_network
 
     @override

--- a/digits/model/images/classification/views.py
+++ b/digits/model/images/classification/views.py
@@ -116,7 +116,8 @@ def create():
                 raise werkzeug.exceptions.BadRequest(
                         'Job not found: %s' % form.previous_networks.data)
 
-            network = fw.get_network_from_previous(old_job.train_task().network)
+            use_same_dataset = (old_job.dataset_id == job.dataset_id)
+            network = fw.get_network_from_previous(old_job.train_task().network, use_same_dataset)
 
             for choice in form.previous_networks.choices:
                 if choice[0] == form.previous_networks.data:

--- a/digits/model/images/generic/views.py
+++ b/digits/model/images/generic/views.py
@@ -101,7 +101,8 @@ def create():
                 raise werkzeug.exceptions.BadRequest(
                         'Job not found: %s' % form.previous_networks.data)
 
-            network = fw.get_network_from_previous(old_job.train_task().network)
+            use_same_dataset = (old_job.dataset_id == job.dataset_id)
+            network = fw.get_network_from_previous(old_job.train_task().network, use_same_dataset)
 
             for choice in form.previous_networks.choices:
                 if choice[0] == form.previous_networks.data:

